### PR TITLE
Migrate to marathon task tag

### PR DESCRIPTION
**Pull request by pbetkier**
Adds a script for migrating registered services to the state expected by #99, i.e. containing marathon-task tag with related marathon task ID.

This script is expected to be executed on the consul cluster with services managed by marathon-consul, in all DCs separately, before updating an existing marathon-consul instance. If consul services are not migrated before update, services managed by marathon-consul will get deregistered and registered again according to the new format during sync at first startup following the update. This will cause ~10 seconds service unavailability in the discovery before service healthchecks become passing.

### CODE REVIEW

**Comment by janisz** [comment]
Would you mind updating Readme please

File: scripts/migrate_to_marathon_task_tag.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+func migrateSingleDC(bootstrapAgentAddress string, consulTag string) *MigrationStats {
+	logrus.WithField("BootstrapAgentAddress", bootstrapAgentAddress).WithField("Tag", consulTag).Info("Starting migration...")
+
+	agentsPort, err := extractPort(bootstrapAgentAddress)
+	if err != nil {
+		logrus.WithError(err).WithField("BootstrapAgentAddress", bootstrapAgentAddress).
+			Error("Could not extract port from agent address, aborting.")
+		os.Exit(1)
```
Comment (janisz) [inline scripts/migrate_to_marathon_task_tag.go]: How about using `Fatal`?

File: scripts/migrate_to_marathon_task_tag.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
```
Comment (janisz) [inline scripts/migrate_to_marathon_task_tag.go]: In other files we alias it with `log`

File: scripts/migrate_to_marathon_task_tag.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+	nodes, _, err := client.Catalog().Nodes(&api.QueryOptions{})
+	if err != nil {
+		logrus.WithError(err).Error("Could not fetch nodes, aborting.")
+		os.Exit(1)
+	}
+
+	logrus.Infof("Discovered %d node(s) to migrate", len(nodes))
+	stats := migrateNodes(nodes, agentsPort, consulTag)
+	logrus.WithField("Stats", stats).Info("Migration finished")
```
Comment (janisz) [inline scripts/migrate_to_marathon_task_tag.go]: How about putting each stat in it's own field

File: scripts/migrate_to_marathon_task_tag.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+			stats.SkippedFailedServices++
+			continue
+		}
+		logrus.WithField("ServiceID", agentService.ID).Info("Migrated service")
+		stats.MigratedServices++
+	}
+}
+
+func extractPort(address string) (int, error) {
+	return strconv.Atoi(address[strings.LastIndex(address, ":")+1:])
```
Comment (janisz) [inline scripts/migrate_to_marathon_task_tag.go]: How do you guarantee `LastIndex` non negative?

File: scripts/migrate_to_marathon_task_tag.go (hunk 3 above)
Comment (pbetkier) [inline scripts/migrate_to_marathon_task_tag.go]: I thought about it, but came to a conclusion that the effect would be the same, with much boilerplate code.

File: scripts/migrate_to_marathon_task_tag.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+			stats.SkippedServicesNotManagedByMarathonConsul++
+			continue
+		}
+		if hasMarathonTaskTag(agentService) {
+			logrus.WithField("ServiceID", agentService.ID).Info("Service already has marathon-task tag, skipping")
+			stats.SkippedServicesAlreadyWithMarathonTaskTag++
+			continue
+		}
+
+		err := nodeClient.Agent().ServiceRegister(migrated(agentService))
```
Comment (medzin) [inline scripts/migrate_to_marathon_task_tag.go]: Maybe ``` if err := nodeClient.Agent().ServiceRegister(migrated(agentService)); err != nil { }```

File: scripts/migrate_to_marathon_task_tag_test.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/allegro/marathon-consul/consul"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/testutil"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMigrateToMarathonTaskTag_shouldMigrateOnlyServicesWithMarathonTagAndWithoutTaskTag(t *testing.T) {
```
Comment (medzin) [inline scripts/migrate_to_marathon_task_tag_test.go]: Maybe you want to use [testify suite](https://godoc.org/github.com/stretchr/testify/suite) here?

File: scripts/migrate_to_marathon_task_tag_test.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	migrateSingleDC(server.HTTPAddr, "marathon")
+
+	// then
+	client, _ := clientToServer(server)
+	clientToOtherDC, _ := clientToServer(serverOnOtherDC)
+	services, _, _ := client.Catalog().Services(&api.QueryOptions{})
+	servicesOnOtherDC, _, _ := clientToOtherDC.Catalog().Services(&api.QueryOptions{})
+	assert.Contains(t, services["serviceA"], "marathon-task:serviceA")
+	assert.NotContains(t, servicesOnOtherDC["serviceB"], "marathon-task:serviceC")
+}
```
Comment (janisz) [inline scripts/migrate_to_marathon_task_tag_test.go]: Missing tests with unreachable consul node.

File: scripts/migrate_to_marathon_task_tag.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+type MigrationStats struct {
+	MigratedServices                          int
+	SkippedFailedNodes                        int
+	SkippedFailedServices                     int
+	SkippedServicesNotManagedByMarathonConsul int
+	SkippedServicesAlreadyWithMarathonTaskTag int
+}
+
+func main() {
```
Comment (janisz) [inline scripts/migrate_to_marathon_task_tag.go]: Could be tested in integration test.
<details>
```go

func TestMainFunc(t *testing.T) {
	t.Parallel()
	// given
	server := consul.CreateConsulTestServer(t)
	defer server.Stop()
	server.AddService("serviceA", "passing", []string{"marathon"})
	os.Args = []string{"./migrate", fmt.Sprintf("--bootstrap-agent-address=%s", server.HTTPAddr)}
	// when
	main()
	//then
	client, _ := clientToServer(server)
	services, _, _ := client.Catalog().Services(&api.QueryOptions{})
	assert.Contains(t, services["serviceA"], "marathon-task:serviceA")
}
```
</details>

File: scripts/migrate_to_marathon_task_tag.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+	config := createConfig()
+	migrateSingleDC(config.BootstrapAgentAddress, config.ConsulTag)
+}
+
+func createConfig() *Config {
+	config := &Config{}
+	flag.StringVar(&config.BootstrapAgentAddress, "bootstrap-agent-address", "127.0.0.1:8500", "Address to one of the agents in the migrated DC")
```
Comment (janisz) [inline scripts/migrate_to_marathon_task_tag.go]: In other places we use `location` instead of `address`

**Comment by janisz** [comment]
Would you mind updating `.travis.yaml` please.

File: README.md [hunk 10]
```diff
@@ -206,6 +208,23 @@ The following section describes known limitations in `marathon-consul`.
...
+an additional tag, `marathon-task`, is added to each registration.
+
+If you update marathon-consul from version 0.x.x to 1.x.x, expect the synchronization phase during the first startup to reregister all services managed by marathon-consul to the new format.
+Depending on the configured healthchecks this may result in many seconds of services unavailability in the registry before their healthchecks become passing. If that's
+not acceptable for you, you should ensure that before updating marathon-consul all managed services contain a `marathon-task:<task-id>` tag, where `<task-id>` is the Marathon task id
+being registered in Consul.
+
+You can run our migration script before updating to 1.x.x to ensure that.
+
+// TODO moar
```
Comment (janisz) [inline README.md]: ?

**Comment by janisz** [comment]
LGTM :+1:

File: README.md (hunk 10 above)
Comment (pbetkier) [inline README.md]: more stuff to come ;)

**Pull request merged**